### PR TITLE
SQLite: Enable ef_compare for decimals

### PR DESCRIPTION
**Pull request by lokalmatador**
Part of #19635

### CODE REVIEW

**Comment by lokalmatador** [comment]
@bricelam, so I went ahead and had a first look at the code (specifically your mentioned [PR](https://github.com/dotnet/efcore/pull/19617) on datetime and the [one](https://github.com/dotnet/efcore/pull/20024) on `ef_mod`) you and then just attempted to follow what I saw there, to somehow get an idea (monkey see, monkey do) of how `ef_compare` could work. The contents of this PR by no means provide a draft or anything close to that. But maybe they provide a basis for discussing the implementation.

What I wanted to ask before actually trying to work in an implementation
- how would I add my own tests for driving my implementation?
- is there some specification (or similar available) guiding test design?

I'll try to have a more thorough look at the code infrastructure next week, yet, any help/pointers would be greatly appreciated - thanks!

**Comment by bricelam** [comment]
It looks like we already have some tests in [test/EFCore.Sqlite.FunctionalTests/BuiltInDataTypesSqliteTest.cs](https://github.com/dotnet/efcore/blob/9a570332dc0de8ef2c7f6a408e023e732133fe94/test/EFCore.Sqlite.FunctionalTests/BuiltInDataTypesSqliteTest.cs#L1116). The comparison is done on the client, but if you enable a server translation, they'll verify that the result is still correct.

File: src/EFCore.Sqlite.Core/Query/Internal/SqliteSqlTranslatingExpressionVisitor.cs [hunk 1]
```diff
@@ -156,6 +153,71 @@ protected override Expression VisitBinary(BinaryExpression binaryExpression)
                         visitedExpression.TypeMapping);
                 }
 
+                if (sqlBinary.OperatorType == ExpressionType.GreaterThan
+                    && (_functionModuloTypes.Contains(GetProviderType(sqlBinary.Left))
+                        || _functionModuloTypes.Contains(GetProviderType(sqlBinary.Right))))
+                {
+                    return SqlExpressionFactory.Function(
+                        "ef_compare_gt",
```
Comment (bricelam) [inline src/EFCore.Sqlite.Core/Query/Internal/SqliteSqlTranslatingExpressionVisitor.cs]: Can have a single `ef_compare` function that returns -1, 0, or 1.

```cs
CreateFunction(
    "ef_compare",
    (decimal? left, decimal? right) => left.HasValue && right.HasValue
        ? decimal.Compare(left.Value, right.Value)
        : default(int?),
    deterministic: true);
```

.NET | SQL
--- | ---
m1 > m2 | ef_compare($m1, $m2) > 0
m1 >= m2 | ef_compare($m1, $m2) >= 0
m1 < m2 | ef_compare($m1, $m2) < 0
m1 <= m2 | ef_compare($m1, $m2) <= 0

Here, the expression here for GreaterThan would be:

```cs
return SqlExpressionFactory.GreaterThan(
    SqlExpressionFactory("ef_compare, ...),
    SqlExpressionFactory.Constant(0));

File: src/EFCore.Sqlite.Core/Query/Internal/SqliteSqlTranslatingExpressionVisitor.cs [hunk 2]
```diff
@@ -156,6 +153,71 @@ protected override Expression VisitBinary(BinaryExpression binaryExpression)
                         visitedExpression.TypeMapping);
                 }
 
+                if (sqlBinary.OperatorType == ExpressionType.GreaterThan
+                    && (_functionModuloTypes.Contains(GetProviderType(sqlBinary.Left))
+                        || _functionModuloTypes.Contains(GetProviderType(sqlBinary.Right))))
```
[the remark this answers is not in the document]
Comment (lokalmatador) [reply src/EFCore.Sqlite.Core/Query/Internal/SqliteSqlTranslatingExpressionVisitor.cs]: Did so, thanks.

File: src/EFCore.Sqlite.Core/Query/Internal/SqliteSqlTranslatingExpressionVisitor.cs (hunk 1 above)
Comment (lokalmatador) [reply src/EFCore.Sqlite.Core/Query/Internal/SqliteSqlTranslatingExpressionVisitor.cs]: Changed it to a single `ef_compare` function

**Comment by lokalmatador** [comment]
Thanks for the tips, I actually now got the tests running and see some errors. Gonna submit an updated over the weekend.

File: src/EFCore.Sqlite.Core/Query/Internal/SqliteSqlTranslatingExpressionVisitor.cs [hunk 3]
```diff
@@ -233,5 +231,32 @@ private static Type GetProviderType(SqlExpression expression)
...
+
+        private Expression DoDecimalCompare(SqlExpression visitedExpression, ExpressionType op, SqlExpression left, SqlExpression right)
+        {
+            var actual = SqlExpressionFactory.Function(
+                name: "ef_compare",
+                arguments: new[] { left, right },
+                nullable: true,
+                argumentsPropagateNullability: new[] { true, true },
+                visitedExpression.Type,
+                visitedExpression.TypeMapping);
```
Comment (bricelam) [inline src/EFCore.Sqlite.Core/Query/Internal/SqliteSqlTranslatingExpressionVisitor.cs]: I think type and type mapping are throwing it off. Try using:
```cs
typeof(int),
null);
```

Comment (bricelam) [reply src/EFCore.Sqlite.Core/Query/Internal/SqliteSqlTranslatingExpressionVisitor.cs]: Well, that fixed one of the test failures.

File: src/EFCore.Sqlite.Core/Query/Internal/SqliteSqlTranslatingExpressionVisitor.cs [hunk 4]
```diff
@@ -233,5 +231,29 @@ private static Type GetProviderType(SqlExpression expression)
                 : (expression.TypeMapping?.Converter?.ProviderClrType
                     ?? expression.TypeMapping?.ClrType
                     ?? expression.Type).UnwrapNullableType();
+
+        private static bool AttemptDecimalCompare(SqlBinaryExpression sqlBinary) =>
+            GetProviderType(sqlBinary.Left) == typeof(decimal)
+            && GetProviderType(sqlBinary.Right) == typeof(decimal);
```
Comment (bricelam) [inline src/EFCore.Sqlite.Core/Query/Internal/SqliteSqlTranslatingExpressionVisitor.cs]: Ah, you're not using `sqlBinary.OperatorType` as a condition here, so it's now translating every decimal operation. Limit this to GreaterThan, GreaterThanOrEqual, LessThan, and LessThanOrEqual to fix the other failures

Comment (lokalmatador) [reply src/EFCore.Sqlite.Core/Query/Internal/SqliteSqlTranslatingExpressionVisitor.cs]: Thanks! Ironically, I had this check before but then refactored it out, thinking that this useless "comparison" is not necessary at this point, as it is anyways done later on (`DoDecimalCompare`). Obviously it is. Now all the tests pass again.

File: src/EFCore.Sqlite.Core/Query/Internal/SqliteSqlTranslatingExpressionVisitor.cs [hunk 5]
```diff
@@ -233,5 +231,33 @@ private static Type GetProviderType(SqlExpression expression)
...
+
+        private Expression DoDecimalCompare(SqlExpression visitedExpression, ExpressionType op, SqlExpression left, SqlExpression right)
+        {
+            var actual = SqlExpressionFactory.Function(
+                name: "ef_compare",
+                new[] { left, right },
+                nullable: true,
+                new[] { true, true },
+                typeof(int));
+            var oracle = SqlExpressionFactory.Constant(value: 0);
```
Comment (smitpatel) [inline src/EFCore.Sqlite.Core/Query/Internal/SqliteSqlTranslatingExpressionVisitor.cs]: what is oracle!?

Comment (AndriySvyryd) [reply src/EFCore.Sqlite.Core/Query/Internal/SqliteSqlTranslatingExpressionVisitor.cs]: oracle: a person or thing regarded as an infallible authority or guide on something.

Comment (bricelam) [reply src/EFCore.Sqlite.Core/Query/Internal/SqliteSqlTranslatingExpressionVisitor.cs]: Kind of like a sentinel but non-arbitrary and with canonical semantics

Comment (lokalmatador) [reply src/EFCore.Sqlite.Core/Query/Internal/SqliteSqlTranslatingExpressionVisitor.cs]: Exactly. The oracle predicts the expected result. But if wanted, I'm happy to rename it to, e.g., `expectedResult` or sth similar.

**Comment by JonPSmith** [comment]
Hi @lokalmatador ,

Can I ask a question about what you are doing? I can't currently filter/order on a decimal in Sqlite database provider. Will your changes fix that?

**Comment by lokalmatador** [comment]
Hi @JonPSmith ,
I'm currently working of various items in this [list](https://github.com/dotnet/efcore/issues/19635). As [mentioned](https://github.com/dotnet/efcore/issues/19635#issuecomment-575948812) by @bricelam (if I got it correctly) , OrderBy won't be implemented for SQLite, yet I can't elaborate on why. For filtering, for now I do not think it's planned, but again, @bricelam probably better knows about that. I eventually just contribute code :smile:

**Comment by JonPSmith** [comment]
Hi @lokalmatador,

Thanks for the update, and for your help in moving EF Core forward. Great stuff. I was asking as I'm working on some docs, so now I know.

**Pull request merged**